Readme.md
# node-falcon

> open-falcon nodejs client

see [open-falcon data push section](http://book.open-falcon.com/zh/usage/data-push.html)

## Installation

    npm install --save open-falcon

## Usage

    'use strict';
    let Falcon = require('open-falcon');
    Falcon.init('http://127.0.0.1:6060');
    // report memory usage every minutes
    let falcon = new Falcon();
    setInterval(function() {
        let usage = process.memoryUsage();
        falcon
            .step(20)
            .tag('type', 'memory')
            .gauge('node.mem.rss', usage.rss)
            .gauge('node.mem.total', usage.heapTotal)
            .gauge('node.mem.used', usage.heapUsed);
    }, 1000 * 60);

## TODO

* Global waiting queue rather than every instance manage it's own metric queue

## License

MIT


History.md
1.1.0 / 2015-10-15

* Support increment type(fake) metric

1.0.0 / 2015-10-14
==================

* Basic falcon initialization and data push


test/index.js
'use strict';
let Falcon = require('..');
let debug = require('debug')('open-falcon:test');

describe('open-falcon', function() {
    this.timeout(5000);

    let api = 'http://localhost:6060';
    let project = 'projectName';
    Falcon.init(api, project);

    it('should initialize api and project', function(done) {
        let falcon = new Falcon({
            step: 0,
        });

        let metric =  'metricName'
        let value = 100;
        falcon.gauge(metric, value);

        getClientRequestBody(6060, function(queue) {
            queue[0].tags.should.equal('project='+project);
            queue[0].metric.should.equal(metric);
            queue[0].value.should.equal(value);
            done();
        });
    });

    it('should support GAUGE and COUNTER counterType', function(done) {
        let falcon = new Falcon({
            step: 0,
        });

        let metric =  'metricName'
        let value = 100;
        falcon.gauge(metric, value);

        getClientRequestBody(6060, function(queue) {
            queue[0].counterType.should.equal('GAUGE');

            falcon.counter(metric, value);
            getClientRequestBody(6060, function(queue) {
                queue[0].counterType.should.equal('COUNTER');
                done();
            });
        });
    });

    it('should support fake INCREMENT counterType', function(done) {
        let falcon = new Falcon();

        let metric = 'metricName';
        falcon
            .step(1)
            .increment(metric)
            .increment(metric);

        setTimeout(function() {
            falcon.increment(metric);
        }, 1000);

        getClientRequestBody(6060, function(queue) {
            queue[0].counterType.should.equal('GAUGE');
            queue[0].value.should.equal(2);

            done();
        });
    });

    it('should tag metric', function(done) {
        let falcon = new Falcon({
            step: 0,
        });

        let metric =  'metricName'
        let value = 100;
        falcon
            .tag('key', 'value')
            .gauge(metric, value);

        getClientRequestBody(6060, function(queue) {
            queue[0].tags.should.equal('project='+project+','+'key=value');
            done();
        });
    });

    it('should allow change step', function(done) {
        let falcon = new Falcon({
            step: 0,
        });

        let metric =  'metricName'
        let value = 100;
        falcon
            .step(1)
            .gauge(metric, value);

        getClientRequestBody(6060, function(queue) {
            queue[0].step.should.equal(1);
            done();
        });
    });
});

let http = require('http');
function getClientRequestBody(port, callback) {
    let server = http.createServer(function(req, res) {
        if(req.method === "POST"){
            let data = '';

            req.on('data', function(chunk){
                data += chunk;
            });
            req.on('end', function(){
                end();
                callback(JSON.parse(data));
            });
        } else {
            end();
        }

        function end() {
            res.end();
            server.close();
        }
    });
    server.listen(port);
}


index.js
'use strict';

let os = require('os');
let http = require('http');
let request = require('request');
let debug = require('debug')('open-falcon:index');

function getTrimedHostname() {
    let hostname = os.hostname();
    if (hostname.split('.').length) {
        return hostname.split('.')[0];
    }
    return hostname;
}

/**
 * var Falcon = require('open-falcon');
 * Falcon.init('http://localhost:6060', 'exampleProject');
 *
 * var falcon1 = new Falcon({
 *   tags: 'tagkey=tagvalue',
 * });
 *
 * falcon1
 *   .step(60)
 *   .tag('anotherkey', 'anothervalue')
 *   .gauge('metricName', 100);
 *
 * var falcon2 = new Falcon();
 * falcon2.counter('metricName2', 20);
 *
 * var falcon3 = new Falcon();
 * falcon3
 *   .step(20)
 *   .increment('incrementMetric');
 */

/**
 * options
 *   step: optional, integer, seconds, default 60
 *   tags: optional
 */
function Falcon(options) {
    debug('constructor', options);
    options = options || {};

    this._lock = false;
    this._queue = [];
    this._increment = null;

    // step trigger autoFlush
    this.step(options.step === undefined ? Falcon.DEFAULT_STEP : options.step);
    this.tag('project', Falcon.PROJECT);
    this._handler = options.handler || Falcon.DEFAULT_HANDLER;
    if (options.tags) {
        this.tag(options.tags);
    }
}

/**
 * api: Falcon push api
 * options
 *   project: project name, set as project=<project> tag
 */
Falcon.DEFAULT_STEP = 60;
Falcon.DEFAULT_COUNTER_TYPE = 'GAUGE';
Falcon.ENDPOINT = getTrimedHostname();

/**
 * @param {string} api
 * @param {string} project project name
 * @param {Function} handler optional request handler
 */
Falcon.init = function(api, project, handler) {
    debug('init', api, project);
    Falcon.API = api;
    Falcon.PROJECT = project;
    Falcon.DEFAULT_HANDLER = handler || function noop(){};
};

Falcon.prototype.tag = function(key, value) {
    debug('tag', key, value);
    if (this._lock) {
        this.emit('error', new Error('Falcon locked, can\'t alter tag'));
        return this;
    }

    if (!this._tags) {
        this._tags = '';
    } else {
        this._tags = this._tags + ',';
    }
    if (arguments.length === 2) {
        this._tags = this._tags + key + '=' + value;
    } else {
        this._tags = this._tags + key;
    }
    return this;
};

Falcon.prototype.step = function(step) {
    debug('step', step);
    if (this._lock) {
        this.emit('error', new Error('Falcon locked, can\'t alter step'));
        return this;
    }

    this._step = step;
    // autoFlush timeout should change everytime step changes
    this.autoFlush();
    return this;
};

Falcon.prototype.gauge = function(metric, value, options) {
    debug('gauge', metric, value, options);
    options = options || {};
    options.counterType = 'GAUGE';
    this.push(metric, value, options);
    return this;
};

Falcon.prototype.counter = function(metric, value, options) {
    debug('counter', metric, value, options);
    options = options || {};
    options.counterType = 'COUNTER';
    this.push(metric, value, options);
    return this;
};

Falcon.prototype.increment = function(metric, options) {
    debug('increment', metric, options);
    options = options || {};
    options.counterType = 'INCREMENT';
    this.push(metric, 1, options);
    return this;
};

/**
 * @private
 */
Falcon.prototype.now = function() {
    return Math.floor(Date.now() / 1000);
};

/**
 * @private
 */
Falcon.prototype.createItem = function(metric, value, options) {
    options.tags = options.tags ? (','+options.tags) : '';

    let endpoint = Falcon.ENDPOINT;
    let timestamp = this.now();
    let step = options.step || this._step;
    let counterType = options.counterType;
    let tags = this._tags + options.tags;

    if (counterType === 'INCREMENT') {
        // reset fake counterType
        counterType = 'GAUGE';
    }

    return {
        metric,
        value,
        endpoint,
        timestamp,
        step,
        counterType,
        tags,
    };
};

/**
 * @private
 */
Falcon.prototype.push = function(metric, value, options) {
    let counterType = options.counterType || Falcon.DEFAULT_COUNTER_TYPE;
    if (counterType === 'GAUGE' || counterType === 'COUNTER') {
        this._queue.push(this.createItem(metric, value, options));
    } else if (counterType === 'INCREMENT') {
        if (!this._increment) {
            this._increment = this.createItem(metric, value, options);
        } else {
            this._increment.value = this._increment.value + 1;
        }
    }
    this.checkAndFlush();
    return this;
};

/**
 * @private
 */
Falcon.prototype.autoFlush = function() {
    debug('autoFlush');
    clearTimeout(this._timeout);
    this.checkAndFlush();

    // if step equals to zero, reset it to 1
    let step = this._step || 1;
    let self = this;
    this._timeout = setTimeout(function() {
        self.autoFlush();
    }, step * 1000);
};

/**
 * @private
 */
Falcon.prototype.checkAndFlush = function() {
    debug('checkAndFlush');
    let timestamp = this.now();
    if (this._increment) {
        if (timestamp - this._increment.timestamp >= this._step) {
            let increment = this._increment;
            this._queue.push(increment);
            // reset increment rather than delete it
            this._increment = this.createItem(increment.metric, 0, {counterType: 'INCREMENT'});
        }
    }

    if (this._queue.length) {
        if (timestamp - this._queue[0].timestamp >= this._step) {
            this.flush();
        }
    }
};

/**
 * @private
 */
Falcon.prototype.flush = function() {
    debug('flush', this._queue);
    let queue = this._queue;
    if (!queue.length) {
        return;
    }
    this._queue = [];

    let self = this;
    request.post({
        url: Falcon.API,
        body: JSON.stringify(queue),
    }, this._handler)
        .on('error', this._handler);
};

module.exports = Falcon;
